### ISSUE TITLE
Comment threads occasionally mismatch with GitHub UI presentation of the change

### ISSUE THREAD

**Issue by joestringer** [ldelossa/gh.nvim#47]
I don't have reproducible steps for this one just yet, but I noticed during a recent review that sometimes the threads don't post against the line that I expect. I created a thread against a line in a specific commit, I think the lines had some diff, then some existing code unchanged, then a couple more lines diff / added. Inside gh.nvim, it displays the comment thread correctly (according to how gh.nvim originally presented it), but on the github UI, the comment is actually posted against a line a couple of lines down.

At least one of the cases, I actually posted the comment against a line _between_ the diff. Maybe the line calculation is slightly off if not commenting directly against the diff...?

I'll keep an eye out and post more details here again if I notice more details.

**Comment by joestringer** [comment]
:thinking: [Looking at the API](https://docs.github.com/en/rest/pulls/comments#create-a-review-comment-for-a-pull-request), it almost seems like `side` is misconfigured in the API request, but I can't figure out exactly why that would be, since the logic seems fairly straightforward.

**Comment by joestringer** [comment]
Ooh, unless it's something to do with github's behaviour for the `position` field...?

> **position** integer
> **This parameter is deprecated. Use `line` instead.** The position in the diff where you want to add a review comment. Note this value is not the same as the line number in the file. For help finding the position value, read the note above.

**Comment by ldelossa** [comment]
Ohh interesting, Unless I just missed it, that deprecation is new. 

What makes you think the side is incorrect?

**Comment by ldelossa** [comment]
It is true, I'm submitting both line and position to the API: https://github.com/ldelossa/gh.nvim/blob/41552a3f7af395aa0d82a0d2ac886f4f172f0245/lua/litee/gh/ghcli/init.lua#L429 

I wonder if it just chooses "line" if both are available, or if this is actually causing an issue.

**Comment by ldelossa** [comment]
Id expect that to happen, since when you reopen gh.nvim, it reloads the threads from the API, so it may have translated it wrong, then read it back from the API. 

I cant seem to reproduce it. You mentioned this was when creating a thread on left side of the diffview?

**Comment by joestringer** [comment]
I think it happened when I opened a thread from the right side of diffview. It must have been somehow related to the specific change though, because I was pretty consistently using `GHCreateThread` from the right view, and most times it worked. It would only occasionally pick a line that's a bit off.

**Comment by ldelossa** [comment]
Hmm, interesting. Ill keep an eye out. It does seem some whacky stuff us going on with creating threads on the left side. See: https://github.com/cli/cli/issues/5746 (tho i could be missing something)

**Comment by joestringer** [comment]
Re-opening as I captured the steps this time:

https://asciinema.org/connect/84d46e39-2ed0-4212-b77b-f28ef127a8fb

It seems related to whether you do a `GHStartReview` and then comment on an earlier commit in the PR, then a latter commit in the PR inserts/removes lines in that file. The comment is posted with a line offset based on the top commit, not the commit that you were commenting on. I noted that the graphql API doesn't seem to have a field for commit, so there's no way we could tell the GitHub API at this point that the thread should be posted against a line offset based on the earlier commit in the PR:

https://github.com/ldelossa/gh.nvim/blob/10d1ae05ce7c4f70dd6594f3e7776527bb15dc57/lua/litee/gh/pr/thread_buffer.lua#L708

https://docs.github.com/en/graphql/reference/input-objects#addpullrequestreviewthreadinput

**Comment by ldelossa** [comment]
Did a little more digging, and this is pretty annoying. 

1) Determined we didn't take into account the "checked out" commit when creating a review. Reviews do take a "commit_id" which is already kinda weird... but w/e lets play their game. So hacked on a few changes which just funnels the currently checked out commit to the Create of the review

```
function M.start_review()
    if s.pull_state == nil then
        lib_notify.notify_popup_with_timeout("Must open a pull request before starting a review.", 7500, "error")
        return
    end
    local commit = nil
    if s.pull_state.last_opened_commit ~= nil then
        commit = s.pull_state.last_opened_commit
    else
        commit = s.pull_state.head
    end
    local review = ghcli.create_review(s.pull_state.number, commit)
    s.pull_state.review = review
    vim.cmd("GHRefreshPR")
end
``` 
(trust that s.pull_state.last_opened_commit will be the last commit opened via the pull request's data tree)

Okay so that creates a review focused on the commit we suspect, and I can confirm that with cli logging, here is the review we receive:

```
[info] [gh] cmd: gh api -X GET -F per_page=100 /repos/{owner}/{repo}/pulls/4/reviews -F page=1 out:
{ {
    _links = {
      html = {
        href = "https://github.com/ldelossa/litee-gh-testing/pull/4#pullrequestreview-1079584031"
      },
      pull_request = {
        href = "https://api.github.com/repos/ldelossa/litee-gh-testing/pulls/4"
      }
    },
    author_association = "OWNER",
    body = "",
    commit_id = "85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9",
    html_url = "https://github.com/ldelossa/litee-gh-testing/pull/4#pullrequestreview-1079584031",
    id = 1079584031,
    node_id = "PRR_kwDOHIASLM5AWSUf",
    pull_request_url = "https://api.github.com/repos/ldelossa/litee-gh-testing/pulls/4",
    state = "PENDING",
```

As you can see, the review has the correct commit_id. 

So now we create a thread via graphql, this request doesn't take a commit ID, instead it take the PR node id. And when this thread. The issue is, when we go and grab the comments within a review thread, it doesn't seem to matter that the review itself is checked out to a particular commit:

```
              node = {
                comments = {
                  edges = { {
                      node = {
                        author = {
                          avatarUrl = "https://avatars.githubusercontent.com/u/5642902?u=4ee288002460604dd7fc551750b0bfdef2c56ef5&v=4",
                          login = "ldelossa",
                          resourcePath = "/ldelossa",
                          url = "https://github.com/ldelossa"
                        },
                        body = "Test\n",
                        commit = {
                          oid = "e6323cbce1fc884a88d47299d3bb070e4fcb6216",
                          parents = {
                            edges = { {
                                node = {
                                  id = "C_kwDOHIASLNoAKDg1Zjk3ZjhiOWUyYjNkNGQyMWM2MDM4MDQ4MmI2YzZkN2NjMDRjZjk",
                                  oid = "85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9"
                                }
                              } }
                          }
                        },
                        createdAt = "2022-08-20T21:26:17Z",
                        id = "PRRC_kwDOHIASLM44qzeT",
                        path = "README.md",
                        position = 5,
                        publishedAt = vim.NIL,
                        pullRequestReview = {
                          id = "PRR_kwDOHIASLM5AWSUf"
                        },
```
This is the comment that is implicitly created by creating a thread with a body. Its pullRequestReview.id property matches the pull request checked out to a particular commit, but the comment references HEAD. 

Very obnoxious, and not sure why this occurs. 

The only ReviewComment mutation which actually takes a commit ID requires a "inReplyTo", which makes one assume it can only be a reply to the implicity comment message created when creating a Thread.

**Comment by ldelossa** [comment]
Okay, if you use graphql to create the review and the thread (including implicit first thread comment) the comment is placed on the right comment. Example:

```
mutation {
  addPullRequestReview(input:{
    pullRequestId: "PR_kwDOHIASLM49bo0a"
    commitOID: "85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9"
    threads: [
      {
      	path: "README.md"
        line: 5
        side: RIGHT
        body: "graphql test"
      }
    ]
  }) {
    clientMutationId
  }
}
```

I came to this conclusion by thinking about the web UI. You don't "start" a review on the web UI, it occurs when you submit a comment and choose "start a review", so i'm assuming behind the scenes its doing this in one atomic operation.

**Comment by ldelossa** [comment]
Alright, I was playing around with github's web UI a bit more and found a couple of funny things. 

1. I believe regardless of whether you open a review on a particular commit, once you create comments those comments are associated with HEAD
2. The Github UI will happily display a comment from a previous commit on the wrong line when viewing HEAD. You can see this demonstrated here with the following series of links
  a. Create a comment @ line:5,HEAD~1 https://github.com/ldelossa/litee-gh-testing/pull/4/commits/85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9#r950769517 (literally click the commits UI and create a comment while looking at that specific commit)
  b. Go to "Changed Files" UI and created comment @ line:5,HEAD https://github.com/ldelossa/litee-gh-testing/pull/4/files#r950770153
  c. Notice on "Changed Files" tab comment created at a. is displayed on the "wrong" line when viewing HEAD https://github.com/ldelossa/litee-gh-testing/pull/4/files#r950769517 
  @joestringer I think this is effectively a reproduction of your bug!
3. Threads provide an "originalLine" field which can be used to determine the line they were originally displayed on. I have no idea how this works when there are multiple parents in the line hierarchy but using this "originalLine" field when its not equal to "line" seems to emulate the behavior we want (and won't display the comment on line 7 when viewing HEAD). 

Maybe this is by design? Maybe when viewing HEAD github UI is trying to say "someone commented on this line, but it has moved here now". This doesn't seem to make much sense to me, but maybe its useful if you do not perform any "commit wise" reviewing. 

Anyway, the change is simple on my end. Will push it up now. Feel free to test.

**Comment by ldelossa** [comment]
Reverted that original fix to handle this better. 

I discovered you **can** obtain the original commit of a comment. This makes everything a lot easier. Working on the fix now: https://github.com/ldelossa/gh.nvim/pull/66

**Comment by rmacklin** [comment]
> 2\. The Github UI will happily display a comment from a previous commit on the wrong line when viewing HEAD.

Can you clarify what you mean by "wrong line"? In my experience, the behavior displayed in GitHub's UI is what I would expect. Specially, if I originally comment on a line in function foo (let's say originally line 42), and then in subsequent commits function bar is added above foo, pushing everything down by 5 lines, then GitHub is "smart" enough to still display my original comment on the same part of function foo which is now line 47, as long as that line is identical to what it was when I commented on line 42. The reason I consider that behavior as expected is that I don't want an unrelated change that wasn't addressing my comment to make my comment "disappear" from the latest version of the branch when it is still relevant. Hope that made sense...

**Comment by ldelossa** [comment]
@rmacklin imagine this comment here: https://github.com/ldelossa/litee-gh-testing/pull/4/files#r950769517 which was created at HEAD~1 was referring to something completely different on line 5 of that commit, then what line 7 has been edited to, since then.

**Comment by rmacklin** [comment]
One thing I think is important to emphasize:

> GitHub is "smart" enough to still display my original comment on the same part of function foo which is now line 47, **as long as that line is identical to what it was when I commented on it** (when it was line 42)

If the line is **not** identical to what it was when the comment was made at the original location, GitHub will **not** show the comment in the diff view for the PR as a whole. Instead, the comment would only show up when viewing the original commit's diff and in the conversation view, where it would be labeled as "Outdated".

**Comment by rmacklin** [comment]
> Re-opening as I captured the steps this time:
> 
> https://asciinema.org/connect/84d46e39-2ed0-4212-b77b-f28ef127a8fb

FYI @joestringer following this link leads to an error "This recorder token belongs to a different user." Could you make the recording public and share the public URL instead?

**Comment by ldelossa** [comment]
> One thing that I think is important to emphasize is this:
> 
> > GitHub is "smart" enough to still display my original comment on the same part of function foo which is now line 47, _**as long as that line is identical to what it was when I commented on it**_ (when it was line 42)
> 
> If the line is _**not**_ identical to what it was when the comment was made at the original location, GitHub will _**not**_ show the comment in the diff view for the PR as a whole. Instead, the comment would only show up when viewing the original commit's diff and in the conversation view, where it would be labeled as "Outdated".
> 
> I hope I understood what you meant by
> 
> > imagine this comment here: https://github.com/ldelossa/litee-gh-testing/pull/4/files#r950769517 which was created at HEAD~1 was referring to something completely different on line 5 of that commit, then what line 7 has been edited to, since then.
> 
> If I misunderstood what you wanted me to imagine, or if the behavior I described is not what you are experiencing, perhaps you could create another example through the web UI to make it clearer.

Isnt this not what is occurring with the link i sent? https://github.com/ldelossa/litee-gh-testing/pull/4/files#r950769517 

That comment was made on the first commit at line 5, but is displayed on the final comment on line 7, and the line **did** change.

**Comment by ldelossa** [comment]
feel free to test the this branch:  #66 

be wary of force pushes, but its close to finished.

**Comment by ldelossa** [comment]
This is mostly fixed. However creating multiline comments within a review will always create a comment on HEAD. I can't seem to figure out the right graphql magic to make that work. 

Opened a discussion here: https://github.com/community/community/discussions/30317

**Comment by joestringer** [comment]
nice! I confirmed that at least for single-line threads this is working correctly with HEAD.

**Comment by ldelossa** [comment]
@joestringer im seeing a little bit of a new issue. 

Rebased commits... they have a new commit hash right. But any comments on on the commits before they were rebased, reference the old hash. This kinda breaks the "commit" panel/view since it doesn't discern that a new commit hash may associate with the old commits comments. 🤔

**Comment by rmacklin** [comment]
I think that GitHub's behavior of "continue to show the comment on any commit in the PR as long as the line is identical to what it was when the comment was first created" means that rebased commits (whose additions/deletions were not actually modified) naturally have comments preserved.

I haven't looked at #66 yet, but are you saying that it didn't quite implement that logic to check "was there an earlier comment added to an identical line? If so, display it"?

**Comment by ldelossa** [comment]
@rmacklin thats how it works when youre viewing the HEAD commit currently. 

but when you "open" a commit in gh.nvim a new panel opens, and we try to map threads based on commit hash there. 

Ill play around with the logic of no longer caring about git sha comparisons and instead doing source line checks.

**Comment by ldelossa** [comment]
https://github.com/ldelossa/gh.nvim/pull/68 

Okay I think have this mostly unwrapped. 

I do use a heuristic to paint a comment *somewhere* when it does not match the source line. Since gh.nvim does not have a "global conversation view" to view threaded messages, like the "conversations" tab in the GitHub UI, we need to actually paint the comment **somewhere**. We do this with a heuristic that does a best guess for the source line, with a preference toward the "line" field of the comment. https://github.com/ldelossa/gh.nvim/pull/68/files#diff-9e38a65108e1245ce5e246b2e8c842900bf438e8dc7cd93457ec6a5911aaecf5R244-R247

**Comment by ldelossa** [comment]
#68 merged with some other improvements, gh.nvim can now show threads for comments which the file was removed from the pull request (not listed via the api anymore) and some fixes to the thread diff preview.

@joestringer - if you come across any issues or inconsistencies moving forward feel free to re-open.
@rmacklin  - thanks for your input here, i don't think I would have considered that github api does source line matching. things seem to work great this way.
